plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs

```diff
@@ -666,17 +712,145 @@ compileExpr e = withContextM 2 (sdToTxt $ "Compiling expr:" GHC.<+> GHC.ppr e) $
+    BooleanCoverage -> do
+      -- Check if the thing we are compiling is a boolean
+      CompileContext {ccNameInfo=nameInfo} <- ask
+      let bool = GHC.getName <$> Map.lookup ''Bool nameInfo -- TODO: This should always return a `Just` and it may be worthwhile to crash the compiler with an error if it doesn't
+          tyHeadName = GHC.getName <$> GHC.tyConAppTyCon_maybe exprType
+          headSymName = GHC.getName <$> findHeadSymbol originalExpr
+          isTrueOrFalse = case headSymName of
+            Nothing -> False
+            -- We have to do it this convoluted way because GHC decides that
+            -- `nm` is a *variable* but `'True` and `'False` are data constructors and
+            -- they are not equal as names even if `originalExpr` is e.g. the expression `True`
+            Just nm ->
+              or [ GHC.getOccString nm == GHC.getOccString con
+                 | Just con <- [ GHC.getName <$> Map.lookup sym nameInfo | sym <- ['True, 'False] ] -- TODO: likewise to above these two should always both return `Just`
+                 , GHC.nameSpacesRelated (GHC.occNameSpace . GHC.getOccName $ nm) (GHC.occNameSpace . GHC.getOccName $ con)
+                 ]
+      if tyHeadName /= bool || isTrueOrFalse
+      then return compiledTerm
+      -- It is a boolean that's not `True` or `False`,
+      -- generate the code:
+      -- ```
+      -- if compiledTerm
+      -- then trace (compiledTerm was true) True
+      -- else trace (compiledTerm was false) False
+      -- ```
+      else do
+        cons <- case GHC.tyConAppTyCon_maybe exprType of
+          Just tc -> getDataCons tc
+          Nothing -> throwSd UnsupportedError $ "PANIC! This error should be unreachable in funtion coverageCompile"
+        (dcTrue, dcFalse) <- case cons of
+          [a, b] -> return (a, b)
+          _      -> throwSd UnsupportedError $ "PANIC! This error should be unreachable in funtion coverageCompile"
+
+        match <- getMatchInstantiated exprType
+        let matched = PIR.Apply () match compiledTerm
+
+        compiledType <- compileTypeNorm exprType
+        resultType <- delayType compiledType
+        let instantiated = PIR.TyInst () matched resultType
+
+        let mkMetadata = CoverageMetadata . foldMap (Set.singleton . ApplicationHeadSymbol . GHC.getOccString)
+        fc <- addBoolCaseToCoverageIndex (toCovLoc src) False (mkMetadata headSymName)
+        falseTerm <- compileExpr (GHC.mkConApp dcFalse [])
+        falseBranch <- delay $ mkTrace compiledType (T.pack . show $ fc) falseTerm
+
+        tc <- addBoolCaseToCoverageIndex (toCovLoc src) True (mkMetadata headSymName)
+        trueTerm <- compileExpr (GHC.mkConApp dcTrue [])
+        trueBranch <- delay $ mkTrace compiledType (T.pack . show $ tc) trueTerm
+        let branches = [trueBranch, falseBranch]
+
+        let applied = PIR.mkIterApp () instantiated branches
+        force applied
+    where
+      findHeadSymbol :: GHC.CoreExpr -> Maybe GHC.Id
```

Comment:
If we're just looking for `True` and `False`, why are we traversing the expression tree at all? Shouldn't it be exactly a variable in that case?

Reply:
This tells us what function returned `True` or `False`. This information is then relayed in the coverage report to be used downstream.

Reply:
Oh I see, this does double duty of providing a diagnostic name for the normal case and identifying `True`/`False` in the degenerate case... a bit obscure IMO. And perhaps only works since you can't apply `True` to things?

Reply:
Wait, I don't see why it would be the case that it only works for that reason?

Reply:
If you're looking for the head symbol and checking whether that is `True,` then you would claim that `True 1 2` is `True`, although it's not. It happens that that's impossible. I just think it's a bit dodgy to say that an _expression_ is true/false if it's _head symbol_ is true or false.

Reply:
Yes but the important thing is that `findHeadSymbol True = True` and `True a b` will never appear in type-correct GHC core right? So this works...

Reply:
Yes, I think it will do the right thing, it's just a bit smelly.

Reply:
Needs an explanation, I guess.

Reply:
I guess we also can't just print the expression itself to go in the note because it's some Core unpleasantness.

Reply:
I honestly don't get how this is smelly. Do we need a note explaining that we get well-typed core as input?

Reply:
I don't mean that to be in any way sarcastic, I just honestly do not understand what the smell is.

Reply:
Taste I guess :) there are two independent things there: what if you want to change what information you put into the string, now that's tied up in this independent check for the structure of the expression.

I'd just have two things: "is it exactly a variable reference to `True` or `False`" (which should be simple now we've cleared up the name issue), and "here's the head symbol for logging"

Reply:
Ok that's fair, I can see a point of organising it that way too, with a separate function `isConstructorTrueOrFalse :: GHC.Expr -> Bool` defined elsewhere.